src/features/decorators.tsx

```diff
@@ -513,23 +571,49 @@ export const toggleDecorations = (flag: boolean) => {
             try {
               const c = Color(`#${r}`);
               const previewIdPrefix = `hex-color-preview-${blockUid}-${r}-`;
+
+              // Filter out spans that do not match the current reference or already have a color preview.
               const renderedRefSpans = renderedRefs.filter(
                 (s) =>
                   s.getAttribute("data-tag") === r &&
                   (!s.lastElementChild ||
                     !s.lastElementChild.id.startsWith(previewIdPrefix))
               );
+
               renderedRefSpans.forEach((renderedRef, i) => {
-                const newSpan = document.createElement("span");
-                newSpan.style.backgroundColor = c.string();
-                newSpan.className = HEX_COLOR_PREVIEW_CLASSNAME;
-                newSpan.id = `${previewIdPrefix}${i}`;
-                renderedRef.appendChild(newSpan);
+                const newInput = document.createElement("input");
+                newInput.type = "color";
+                newInput.value = c.hex();
+                if (newInput.value.toLocaleLowerCase() === "#ffffff")
+                  newInput.style.boxShadow = "inset 0 0 0 1px hsla(0,0%,0%,.2)";
+                newInput.className = HEX_COLOR_PREVIEW_CLASSNAME;
+                newInput.id = `${previewIdPrefix}${i}`;
+                newInput.onmousedown = (e) => e.stopPropagation();
+                renderedRef.appendChild(newInput);
+                newInput.onchange = async (e) => {
+                  const target = e.target as HTMLInputElement;
+                  const newColor = target.value;
+                  const uid = getBlockUidFromTarget(target);
+                  const blockText = getTextByBlockUid(uid);
+                  const newText = blockText.replace(
+                    new RegExp(`#${r}`, "g"),
+                    newColor
+                  );
+                  await window.roamAlphaAPI.updateBlock({
+                    block: { uid, string: newText },
+                  });
+                  target.value = newColor;
+                  toggleDecorations(false);
+                  setTimeout(() => {
+                    toggleDecorations(true);
+                  }, 100);
```

Comment:
is there a better way to do this?  This causes a flicker because it removes the `css`.